src/ape/api/query.py

```diff
@@ -0,0 +1,100 @@
+from typing import Any, Dict, List, Literal, Optional, Union
+
+import pandas as pd
+from pydantic import BaseModel, PositiveInt, root_validator
+
+from ape.managers.networks import NetworkManager
+from ape.types import AddressType
+from ape.utils import abstractdataclass, abstractmethod
+
+QueryType = Union["BlockQuery", "AccountQuery", "ContractEventQuery", "ContractMethodQuery"]
+
+
+class _BaseQuery(BaseModel):
+    type: str  # Used as discriminator
+    columns: List[str]
+
+
+class _BaseBlockQuery(_BaseQuery):
+    start_block: PositiveInt = 0
+    stop_block: Optional[PositiveInt] = None
+
+    @root_validator(pre=True)
+    def check_start_block_before_stop_block(cls, values):
+        if values["stop_block"] < values["start_block"]:
+            raise ValueError(
+                f"stop_block: {values['stop_block']} cannot be less than "
+                f"start_block: {values['start_block']}."
+            )
+
+        return values
+
+
+class BlockQuery(_BaseBlockQuery):
+    type: Literal["blocks"]
+
+
+class _BaseAccountQuery(BaseModel):
+    start_nonce: PositiveInt = 0
+    stop_nonce: Optional[PositiveInt] = None
+
+    @root_validator(pre=True)
+    def check_start_nonce_before_stop_nonce(cls, values):
+        if values["stop_nonce"] < values["start_nonce"]:
```

Comment:
Since `stop_nonce` is optional, are we potentially comparing an `int` with `None` here?

Reply:
Yeah, it will require the same treatment as we do the BlockContainer query method, except on the account:

```python
        if stop_block is None:
            stop_block = len(self)

        if stop_block > len(self):
            raise ChainError(
                f"'stop_block={stop_block}' cannot be greater than the chain length ({len(self)}). "
                f"Use '{self.poll_blocks.__name__}()' to wait for future blocks."
            )
```